File: golang-devildyw/simple/L35搜索插入位置.go
package main

import "fmt"

func main() {

	nums := []int{1, 3, 5, 6}
	target := 5
	result := searchInsert(nums, target)
	fmt.Println(result)
}

func searchInsert(nums []int, target int) int {
	var left = 0
	var right = len(nums) - 1
	var result = len(nums)
	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] >= target {
			result = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}

	return result
}

func searchInsertNew(nums []int, target int) int {
	var left = 0
	var right = len(nums) - 1
	var result = len(nums)
	for left <= right {
		mid := left + (right-left)/2
		if nums[mid] >= target {
			// target存在时 mid 等于他的下标 target不存在时 在大于target的第一个元素的位置上
			result = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}

	return result
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L35搜索插入位置.java
package top.dyw.leetcode.Hot100;

public class L35搜索插入位置 {
    public int searchInsert(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        int result = 0;
        while(left <= right) {
            int mid = left + (right - left) / 2;
            if (nums[mid] == target) {
                return mid;
            }
            if (nums[mid] > target) {
                right = mid - 1;
            } else {
                left = mid + 1;
                result = left;
            }
        }
        return result;
    }
}
